File: app/src/main/java/com/merino/ddfilms/ui/fragment/WatchlistFragment.java
package com.merino.ddfilms.ui.fragment;

import static com.merino.ddfilms.utils.StringUtils.WATCH_LIST;
import static com.merino.ddfilms.utils.Utils.showMessage;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.merino.ddfilms.R;
import com.merino.ddfilms.adapters.MovieAdapter;
import com.merino.ddfilms.api.FirebaseManager;
import com.merino.ddfilms.model.Movie;
import com.merino.ddfilms.ui.MainActivity;
import com.merino.ddfilms.ui.SearchActivity;
import com.merino.ddfilms.ui.components.Fab.FabHost;
import com.merino.ddfilms.ui.components.Fab.ShowsFab;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class WatchlistFragment extends Fragment implements FabHost, ShowsFab {

    private MovieAdapter movieAdapter;
    private RecyclerView movieListRecyclerView;
    private boolean isEditMode = false;
    private Toolbar toolbar;
    private MenuItem doneMenuItem;
    private MenuItem moreActionsMenuItem;
    private final FirebaseManager firebaseManager = new FirebaseManager();
    private List<Movie> movieList = new ArrayList<>();
    private final String userID = firebaseManager.getCurrentUserUID();

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_watchlist, container, false);

        setupViews(view);
//        setupToolbar();
        setupRecyclerView();
        loadMoviesFromList();

        return view;
    }

    private void setupViews(View view) {
        movieListRecyclerView = view.findViewById(R.id.movie_list_recycler_view);
//        toolbar = view.findViewById(R.id.toolbar);
    }

    private void setupToolbar() {
        AppCompatActivity activity = (AppCompatActivity) getActivity();
        if (activity == null) return;

        activity.setSupportActionBar(toolbar);
        ActionBar bar = activity.getSupportActionBar();
        if (bar != null) {
            bar.setDisplayHomeAsUpEnabled(true);
            bar.setTitle("Pendientes");
        }

        // Opcional: manejar clic en flecha de regreso
        toolbar.setNavigationOnClickListener(v -> activity.onBackPressed());
    }

    private void setupAddMovieFragment() {
        Intent intent = new Intent(getContext(), SearchActivity.class);
        intent.putExtra("collection", WATCH_LIST);
        intent.putExtra("documentID", firebaseManager.getCurrentUserUID());
        intent.putExtra("listName", "Pendientes");

        int[] moviesID = movieList.stream()
                .mapToInt(Movie::getId)
                .toArray();
        intent.putExtra("moviesID", moviesID);
        startActivity(intent);
    }

    @SuppressLint("NonConstantResourceId")
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                if (isEditMode) {
                    exitEditMode();
                } else {
//                    onBackPressed();
                }
                return true;
            case R.id.action_more:
//                showBottomSheet();
                return true;
            case R.id.action_done:
                exitEditMode();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void setupRecyclerView() {
        movieAdapter = new MovieAdapter();
        movieAdapter.setOnItemLongClickListener(position -> {
            if (!isEditMode) {
                enterEditMode();
            }
            return true;
        });

        movieAdapter.setOnDeleteClickListener((position, movie) -> {
            firebaseManager.deleteMovieFromList(userID, movie, (result, error) -> {
                if (error != null) {
                    showMessage(getContext(), error.getMessage());
                } else if (result != null) {
                    movieAdapter.removeMovie(position);
                    movieList.remove(movie);
                }
            });
        });
        movieListRecyclerView.setAdapter(movieAdapter);
        movieListRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
    }

    @SuppressLint("NotifyDataSetChanged")
    private void updateMoviesList(List<Movie> movies) {
        // Ponemos las películas agregadas más recientemente primero
        Collections.reverse(movies);
        movieAdapter.setMovies(movies);
        movieAdapter.notifyDataSetChanged();
        movieList = movies;
    }

    private void loadMoviesFromList() {
        firebaseManager.loadMovieFromList(userID, WATCH_LIST, (movies, error) -> {
            if (error != null) {
                showMessage(getContext(), error.getMessage());
            } else if (movies != null) {
                updateMoviesList(movies);
            }
        });
    }

    private void enterEditMode() {
        isEditMode = true;
        doneMenuItem.setVisible(true);
        moreActionsMenuItem.setVisible(false);
        movieAdapter.setEditMode(true);
        movieAdapter.setListID(userID);
        ((MainActivity) getActivity()).setFabVisibility(false);
    }

    private void exitEditMode() {
        isEditMode = false;
        doneMenuItem.setVisible(false);
        moreActionsMenuItem.setVisible(true);
        movieAdapter.setEditMode(false);
        ((MainActivity) getActivity()).setFabVisibility(true);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        movieAdapter.setEditMode(false);
    }

    @Override
    public void onResume() {
        super.onResume();
        loadMoviesFromList();
    }

    @Override
    public void onFabClicked() {
        setupAddMovieFragment();
    }
}

File: app/src/main/java/com/merino/ddfilms/ui/fragment/WatchlistFragment.kt
package com.merino.ddfilms.ui.fragment

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.itemsIndexed
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.fragment.app.Fragment
import com.merino.ddfilms.R
import com.merino.ddfilms.api.FirebaseManager
import com.merino.ddfilms.model.Movie
import com.merino.ddfilms.ui.MainActivity
import com.merino.ddfilms.ui.MovieDetailActivity
import com.merino.ddfilms.ui.SearchActivity
import com.merino.ddfilms.ui.components.Fab.FabHost
import com.merino.ddfilms.ui.components.Fab.ShowsFab
import com.merino.ddfilms.ui.components.MoviePosterCard
import com.merino.ddfilms.ui.theme.CinematicTheme
import com.merino.ddfilms.utils.StringUtils.WATCH_LIST
import com.merino.ddfilms.utils.Utils.showMessage

class WatchlistFragment : Fragment(), FabHost, ShowsFab {

    private val firebaseManager = FirebaseManager.getInstance()
    private val movieListState = mutableStateOf<List<Movie>>(emptyList())
    private val isLoadingState = mutableStateOf(true)
    private val isEditModeState = mutableStateOf(false)
    private var userID: String? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        userID = firebaseManager.getCurrentUserUID()

        return ComposeView(requireContext()).apply {
            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
            setContent {
                CinematicTheme {
                    WatchlistScreen(
                        moviesList = movieListState.value,
                        isLoading = isLoadingState.value,
                        isEditMode = isEditModeState.value,
                        onEditModeToggle = {
                            isEditModeState.value = !isEditModeState.value
                        },
                        onMovieClick = { movie, view ->
                            val intent = Intent(requireContext(), MovieDetailActivity::class.java).apply {
                                putExtra("movie", movie)
                            }
                            if (view != null && activity != null) {
                                val options = androidx.core.app.ActivityOptionsCompat.makeSceneTransitionAnimation(
                                    requireActivity(),
                                    view,
                                    "moviePosterTransition"
                                )
                                androidx.core.app.ActivityCompat.startActivity(requireContext(), intent, options.toBundle())
                            } else {
                                startActivity(intent)
                            }
                        },
                        onDeleteMovie = { index, movie ->
                            deleteMovie(index, movie)
                        },
                        onBackClick = {
                            if (parentFragmentManager.backStackEntryCount > 0) {
                                parentFragmentManager.popBackStack()
                            } else {
                                (activity as? MainActivity)?.loadFragment(ProfileFragment())
                            }
                        }
                    )
                }
            }
        }
    }

    private fun loadMoviesFromList(showLoading: Boolean = true) {
        val uid = userID ?: return
        if (showLoading && movieListState.value.isEmpty()) {
            isLoadingState.value = true
        }
        firebaseManager.getMoviesFromList(uid, WATCH_LIST) { movies, error ->
            isLoadingState.value = false
            if (error != null) {
                showMessage(context, error.message)
            } else if (movies != null) {
                movieListState.value = movies.reversed()
            }
        }
    }

    private fun deleteMovie(index: Int, movie: Movie) {
        val uid = userID ?: return
        // Optimistic UI update: remove item instantly from local state
        val current = movieListState.value.toMutableList()
        if (index in current.indices) {
            current.removeAt(index)
            movieListState.value = current
        }

        firebaseManager.deleteMovieFromList(WATCH_LIST, uid, movie) { _, error ->
            if (error != null) {
                showMessage(context, error.message)
                // Revert state if remote deletion fails
                loadMoviesFromList(showLoading = false)
            }
        }
    }

    private fun setupAddMovieFragment() {
        val intent = Intent(context, SearchActivity::class.java).apply {
            putExtra("collection", WATCH_LIST)
            putExtra("documentID", userID)
            putExtra("listName", "Pendientes")
            val moviesID = movieListState.value.map { it.id }.toIntArray()
            putExtra("moviesID", moviesID)
        }
        startActivity(intent)
    }

    override fun onResume() {
        super.onResume()
        loadMoviesFromList()
    }

    override fun onFabClicked() {
        setupAddMovieFragment()
    }
}

@Composable
fun WatchlistScreen(
    moviesList: List<Movie>,
    isLoading: Boolean,
    isEditMode: Boolean,
    onEditModeToggle: () -> Unit,
    onMovieClick: (Movie, View?) -> Unit,
    onDeleteMovie: (Int, Movie) -> Unit,
    onBackClick: () -> Unit
) {
    Surface(
        modifier = Modifier.fillMaxSize(),
        color = MaterialTheme.colorScheme.background
    ) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .statusBarsPadding()
        ) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 16.dp, vertical = 12.dp)
            ) {
                IconButton(onClick = onBackClick) {
                    Icon(
                        painter = painterResource(id = R.drawable.ic_arrow_back),
                        contentDescription = "Volver",
                        tint = MaterialTheme.colorScheme.onBackground
                    )
                }
                Spacer(modifier = Modifier.width(8.dp))
                Text(
                    text = "Películas Pendientes",
                    style = MaterialTheme.typography.titleLarge.copy(fontWeight = FontWeight.Bold),
                    modifier = Modifier.weight(1f)
                )

                IconButton(onClick = onEditModeToggle) {
                    Icon(
                        painter = painterResource(
                            id = if (isEditMode) R.drawable.ic_done else R.drawable.ic_edit
                        ),
                        contentDescription = "Editar",
                        tint = MaterialTheme.colorScheme.primary
                    )
                }
            }

            if (isLoading) {
                Box(
                    modifier = Modifier.fillMaxSize(),
                    contentAlignment = Alignment.Center
                ) {
                    CircularProgressIndicator(color = MaterialTheme.colorScheme.primary)
                }
            } else if (moviesList.isEmpty()) {
                Box(
                    modifier = Modifier.fillMaxSize(),
                    contentAlignment = Alignment.Center
                ) {
                    Text(
                        text = "Tu watchlist está vacía.",
                        color = MaterialTheme.colorScheme.onBackground.copy(alpha = 0.6f),
                        fontSize = 16.sp
                    )
                }
            } else {
                LazyVerticalGrid(
                    columns = GridCells.Fixed(3),
                    contentPadding = PaddingValues(start = 16.dp, end = 16.dp, bottom = 80.dp),
                    horizontalArrangement = Arrangement.spacedBy(12.dp),
                    verticalArrangement = Arrangement.spacedBy(12.dp),
                    modifier = Modifier.fillMaxSize()
                ) {
                    itemsIndexed(
                        moviesList,
                        key = { index, movie -> "${movie.id}_${movie.createdAt ?: ""}_${movie.title ?: ""}_$index" }
                    ) { index, movie ->
                        Box(modifier = Modifier.fillMaxWidth()) {
                            MoviePosterCard(
                                movie = movie,
                                onClick = { view -> onMovieClick(movie, view) },
                                modifier = Modifier.fillMaxWidth()
                            )

                            if (isEditMode) {
                                IconButton(
                                    onClick = { onDeleteMovie(index, movie) },
                                    modifier = Modifier
                                        .align(Alignment.TopEnd)
                                        .padding(4.dp)
                                        .size(28.dp)
                                        .background(
                                            color = Color(0xFFBA1A1A),
                                            shape = RoundedCornerShape(14.dp)
                                        )
                                ) {
                                    Icon(
                                        painter = painterResource(id = R.drawable.ic_remove),
                                        contentDescription = "Eliminar",
                                        tint = Color.White,
                                        modifier = Modifier.size(16.dp)
                                    )
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}
